docs/source/format/CProtocol.rst

```diff
@@ -0,0 +1,264 @@
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   MUST be a JSON-compatible mapping of ASCII strings to ASCII strings.
+   Whitespace MUST only use the 0x20 character. Example::
+
+      {"key1": "base64-encoded value1", "key2": "base64-encode value2"}
+
+   If omitted, MUST be NULL (not an empty string).
+
+   Consumers MAY choose to ignore metadata.  Even then, they SHOULD keep
+   the metadata string around so that they can propagate the information
+   to other consumers.
+
+.. c:member:: int64_t ArrowArray.length
+
+   Mandatory.  The logical length of the array (i.e. its number of items).
+
+.. c:member:: int64_t ArrowArray.null_count
+
+   Mandatory.  The number of null items in the array.  MAY be -1 if not
+   yet computed.
+
+.. c:member:: int64_t ArrowArray.offset
+
+   Mandatory.  The logical offset inside the array (i.e. the number of items
+   from the physical start of the buffers).  MUST be 0 or positive.
+
+   Producers MAY specify that they will only produce 0-offset arrays to
+   ease implementation of consumer code.
+   Consumers MAY decide not to support non-0-offset arrays, but they
+   should document this limitation.
+
+.. c:member:: int64_t ArrowArray.n_buffers
+
+   Mandatory.  The number of physical buffers backing this array.  The
+   number of buffers is a function of the data type, as described in the
+   :ref:`Columnar format specification <format_columnar>`.
+
+   Buffers of children arrays are not included.
+
+.. c:member:: const void** ArrowArray.buffers
+
+   Mandatory.  A C array of pointers to the start of each physical buffer
+   backing this array.  Each `void*` pointer is the physical start of
+   a contiguous buffer.  There must be :c:member:`ArrowArray.n_buffers` pointers.
+
+   The producer MUST ensure that each contiguous buffer is large enough to
+   represent `length + offset` values encoded according to the
+   :ref:`Columnar format specification <format_columnar>`.
+
+   The pointer to the null bitmap buffer, if the data type specifies one,
+   MAY be NULL only if :c:member:`ArrowArray.null_count` is 0.
+
+   Buffers of children arrays are not included.
+
+.. c:member:: int64_t ArrowArray.n_children
+
+   Mandatory.  The number of children this array has.  The number of children
+   is a function of the data type, as described in the
+   :ref:`Columnar format specification <format_columnar>`.
+
+.. c:member:: ArrowArray** ArrowArray.children
+
+   Optional.  A C array of pointers to each child array of this array.
+   There must be :c:member:`ArrowArray.n_children` pointers.
+
+   MAY be NULL only if :c:member:`ArrowArray.n_children` is 0.
+
+.. c:member:: ArrowArray* ArrowArray.dictionary
+
+   Optional.  A pointer to the underlying array of dictionary values.
+
+   MUST be present if the format string indicates a dictionary type.
+   MUST be NULL otherwise.
+
+.. c:member:: void (*ArrowArray.release)(struct ArrowArray*)
```

Comment:
Making this optional seems unnecessary. We could simplify release semantics by simply noting that producers are free to make the release callback a no-op. Then let a released array be signaled with null release callback instead of null format

Reply:
Perhaps. Though producing a C callback may involve non-trivial boilerplate in some languages, which is why I thought being able to pass null may be convenient.